### PR TITLE
Optimize getNoteRole to single correlated-subquery round trip

### PR DESCRIPTION
**Pull request by otomatty**
## 概要

`getNoteRole` を複数の独立クエリから単一の相関サブクエリベースのクエリに最適化し、DB 往復を 3 回から 1 回に削減します。同時に、複数の React Context で value オブジェクトをメモ化し、不要な再レンダリングを防ぎます。

### API サーバー (`server/api/src`)

- **`routes/notes/helpers.ts`**: `getNoteRole` を再実装
  - note 行・メンバーロール・ドメインロールを 1 クエリ（相関サブクエリ）で取得
  - メンバーロール取得を相関サブクエリに統合（case-insensitive email マッチ）
  - ドメインロール取得を相関サブクエリに統合し、SQL 側で最強ロール（editor > viewer）を集約
  - 優先順位（owner > member > domain > visibility > none）を JS 側で適用
  - `getTableColumns` を使用して note 列をフラットに展開

- **`__tests__/routes/notes/helpers.test.ts`**: テストを新クエリ形式に対応
  - `mockRoleRow()` ヘルパーを追加（combined 行を組み立て）
  - 各テストを単一 DB スロット消費に更新
  - 「複数ドメインルール」「大文字小文字マッチ」テストを削除（SQL 側で処理済み）
  - 「unlisted visibility」テストを追加

- **`services/lintEngine/index.ts`**: 防御的上限を追加
  - `DEFAULT_LINT_FINDINGS_LIMIT = 500` を定義
  - `getUnresolvedFindings` と `getFindingsForPage` に `.limit()` を適用
  - バックログが膨大なユーザーでも無制限の結果セットをメモリに読み込まない安全弁

- **`__tests__/services/lintFindings.test.ts`**: 新規テスト
  - findings クエリが防御的 LIMIT を適用することを検証

- **`__tests__/routes/notes/search.test.ts`** / **`__tests__/routes/pageSnapshots.test.ts`** / **`__tests__/services/pageAccessService.test.ts`**: モック DB スロットを更新
  - `getNoteRole` が 1 クエリで note + memberRole/domainRole を返すように調整

### Hocuspocus サーバー (`server/hocuspocus/src`)

- **`coalescePerKey.ts`**: 新規ユーティリティ
  - 非同期 worker をキー単位でコアレッシングするラッパー
  - in-flight 中の重複呼び出しを「追走フラグ」に集約
  - バースト（連続保存ごとのグラフ再構築 HTTP）を間引く

- **`coalescePerKey.test.ts`**: 新規テスト
  - 同一キーの多重起動防止
  - 異なるキーの並行実行
  - trailing 実行の集約
  - エラーハンドリング

- **`index.ts`**: グラフ同期トリガを最適化
  - `triggerGraphSyncCoalesced` を追加（`coalescePerKey` でラップ）
  - `saveDocumentToDb` から呼び出し

### React フロントエンド (`src`)

- **`contexts

https://claude.ai/code/session_01XgbeNj9c8fk6Tyg1JgioDU

## Summary by CodeRabbit

* **Performance Improvements**
  * Single-step note access resolution; lint findings queries capped at 500; per-page graph-sync coalescing; debounced editor change events; memoized contexts/selectors to reduce re-renders.

* **Bug Fixes**
  * Tightened access-role precedence and visibility behavior, including email-related edge-case handling for private notes and consistent search/snapshot access checks.

* **Tests**
  * Updated and added tests covering role-resolution, search/snapshot access, lint-findings limits, and per-key coalescing.

[Review Change Stack](https://app.coderabbit.ai/change-stack/otomatty/zedi/pull/997?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: server/api/src/routes/notes/helpers.ts [hunk 1]
```diff
@@ -169,70 +169,65 @@ export async function getNoteRole(
...
+  // email / domain が空ならサブクエリは何にも一致せず null を返す。
+  // Members match case-insensitively (covers legacy/manual rows). When the
+  // email/domain is empty the subqueries match nothing and yield null.
+  const rows = await db
+    .select({
+      ...getTableColumns(notes),
+      memberRole: sql<NoteMemberRole | null>`(
+        SELECT ${noteMembers.role} FROM ${noteMembers}
+        WHERE ${noteMembers.noteId} = ${noteId}
+          AND LOWER(${noteMembers.memberEmail}) = ${normalizedEmail}
```
[the remark this answers is not in the document]
Comment (otomatty) [reply server/api/src/routes/notes/helpers.ts:202]: ご指摘のとおりリグレッションでした。fdb25a0 で修正しました。

- `normalizedEmail` が空のときは member サブクエリを実行せず `NULL` を select するようにし、空メール行（`member_email = ''`）への偶発一致で匿名アクセスが付与される経路を塞ぎました（domain も同様）。
- 加えて JS 側でも email/domain 不在時はロールを採用しない多層防御を入れています。
- email 無しの呼び出しで member ロールを付与しないことを検証する回帰テストを追加しました。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/api/src/routes/notes/helpers.ts [hunk 2]
```diff
@@ -169,70 +169,65 @@ export async function getNoteRole(
...
+        END
+        FROM ${noteDomainAccess}
+        WHERE ${noteDomainAccess.noteId} = ${noteId}
+          AND ${noteDomainAccess.domain} = ${domain}
+          AND ${noteDomainAccess.isDeleted} = false
+      )`,
+    })
+    .from(notes)
+    .where(and(eq(notes.id, noteId), eq(notes.isDeleted, false)))
+    .limit(1);
```
[the remark this answers is not in the document]
Comment (otomatty) [reply server/api/src/routes/notes/helpers.ts]: ご提案どおり、`normalizedEmail` が空 / `domain` が `null` の場合はサブクエリを実行せず `NULL` を select するよう fdb25a0 で対応しました。不要なサブクエリ実行の回避に加え、空メール行への偶発一致による権限付与（別スレッドのセキュリティ指摘）の防止も兼ねています。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/hocuspocus/src/coalescePerKey.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    if (pending.has(key)) {
+      pending.delete(key);
+      // 最新状態を取り込むためにもう一度だけ実行する。
+      await run(key);
+    }
+  }
+
+  return run;
+}
```
[the remark this answers is not in the document]
Comment (otomatty) [reply server/hocuspocus/src/coalescePerKey.ts:50]: 検討しました。本ヘルパーは単一箇所の fire-and-forget 用途専用（`void triggerGraphSyncCoalesced(pageId).catch(...)`）であり、ご指摘の各点は現状の使用では実害がないため、シンプルな現行実装を維持します。

1. **解決済み Promise の返却**: 戻り値は `void` で破棄され、グラフ同期の完了を待つ下流処理はありません。in-flight 時に即解決の Promise を返しても呼び出し側の挙動に影響しません。
2. **エラーの握りつぶし**: worker (`triggerGraphSync`) は内部で fetch エラー / タイムアウトを捕捉して解決し throw しません。コアレッサでの握りつぶしは意図的で、呼び出し側 `.catch` は防御的な保険です。
3. **再帰によるスタック消費**: 追走は `await run(key)` で都度マイクロタスク境界を挟むため同期コールスタックは伸びず、キー毎に「実行中 1 + 待機 1」に収束します。

ご提案の promise 追跡方式は汎用ユーティリティとしては堅牢ですが、現状の単一用途には過剰（最小実装を優先する方針）と判断しました。他箇所で汎用的に再利用する段になれば、その方式への移行を検討します。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/hocuspocus/src/coalescePerKey.test.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      .fn<(key: string) => Promise<void>>()
+      .mockRejectedValueOnce(new Error("boom"))
+      .mockResolvedValue(undefined);
+    const run = coalescePerKey(worker);
+
+    await run("page-1");
+    await run("page-1");
+
+    expect(worker).toHaveBeenCalledTimes(2);
+  });
```
[the remark this answers is not in the document]
Comment (otomatty) [reply server/hocuspocus/src/coalescePerKey.test.ts:75]: 本体スレッド（coalescePerKey）のとおり、エラーを意図的に握りつぶす（呼び出し側へ伝播させない）設計を維持するため、本テストは「worker が reject しても以降の実行を阻害しない」挙動の確認として現状のままとします。

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -203,9 +208,19 @@ jobs:
...
+      # Install Playwright's browser binaries plus the matching OS deps. Guard
+      # against network stalls: bound each attempt with `timeout` (so a hung
+      # download exits non-zero and triggers a retry) and retry up to 3 times
+      # via the same wretry.action used for checkout above.
+      - name: Install Playwright browsers (with retry)
+        uses: Wandalen/wretry.action@v3
+        with:
+          command: timeout 300 bunx playwright install --with-deps chromium
+          attempt_limit: 3
+          attempt_delay: 10000
```
[the remark this answers is not in the document]
Comment (otomatty) [reply .github/workflows/ci.yml:223]: SHA 固定の方針自体には同意です。ただ本 PR では見送り、既存の専用 issue #975（`Wandalen/wretry.action` をリポジトリ全体で SHA 固定）で一括対応とさせてください。理由:

1. **一貫性**: 本ファイルには既に 13 箇所の `Wandalen/wretry.action@v3` があり、今回追加した 1 箇所だけを固定すると tag / SHA が混在します。#975 で全 workflow + `.github/actions/setup-toolchain/action.yml` をまとめて固定する方が整合的です。
2. **SHA の検証**: 供給チェーン固定は、SHA が `v3` を正しく指すことを検証した上で行うべきで、これは #975 のスコープで実施します。
3. **ブロッキングではない**: zizmor の unpinned-uses は既存の全ジョブにも該当し、現状 main を含め CI の必須チェックにはなっていません。今回の追加は既存の慣習（`@v3`）に揃えたものです。

本 PR の趣旨（パフォーマンス改善 + E2E のハング対策）からも外れるため、固定は #975 に委ねます。

---
_Generated by [Claude Code](https://claude.ai/code)_
